### ISSUE TITLE
Segmentation mask probelm

### ISSUE THREAD

**Issue by rorro6787** [ultralytics/ultralytics#16925]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have trained a YOLO model for segmentation task and I obtained bad results in the validation process so I started researching what could go wrong. I started doing some tests with some pictures and I got weird results. This is the code I use for testing:
```python
def try_YOLO(image_name: str, output_path: str, model="yolov8n-seg-me.pt") -> str:
    model_path = os.path.join(os.getcwd(), model)
    model = YOLO(model=model_path, task="segment", verbose=False)
    img = os.path.join(image_name, image_name)
    results = model(
        img, save=True, project=output_path, verbose=True, show_boxes=False
    )

    masks = results[0].masks
    print(masks.data)
```
When I run that script with an input image, I get this is on the console indicating that there was segmentation detected:
```python
test/images/P41_T1_FLAIR_coronal_80.png: 320x320 1 multiple_esclerosis, 6.9ms
```
However, when I print the mask tensor, all I get is full zero tensor what tells me that there should not be any segmentation:

```python
tensor([[[0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         ...,
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]]], device='cuda:0')
```
This is really problematic because the same issue that is making the model say that there was segmentation when there was not, is making my validation results terrible.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rorro6787, thank you for reaching out to Ultralytics 🚀!

It sounds like you're experiencing an issue with segmentation mask outputs. To help us investigate further, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/)? This will greatly aid in diagnosing the problem.

For initial troubleshooting, please make sure you are using the latest version of the `ultralytics` package. You can upgrade it using the command below:

```bash
pip install -U ultralytics
```

We also recommend visiting our [Docs](https://docs.ultralytics.com) for more details on segmentation tasks and to ensure all settings align with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For further support, you can join the Ultralytics community on [Discord](https://ultralytics.com/discord) 🎧 for real-time chat or explore more in-depth discussions on [Discourse](https://community.ultralytics.com). The [Subreddit](https://reddit.com/r/ultralytics) is also a great place to share experiences.

This is an automated response, but rest assured an Ultralytics engineer will assist you soon. 😊

**Comment by Y-T-G** [comment]
It's not all zeros. There are some values that are one. It corresponds to the pixels that contain the segment.

You can run `print(masks.data.max())` to verify the max value.

**Comment by rorro6787** [comment]
That's the thing, even though it detects sclerosis:
```python
test/images/P41_T1_FLAIR_coronal_80.png: 320x320 1 multiple_esclerosis, 7.2ms
```
This is the max value detected:
```python
tensor(0., device='cuda:0')
```

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by rorro6787** [comment]
```python
Ultralytics 8.3.7 🚀 Python-3.12.3 torch-2.3.1+cu121 CUDA:0 (NVIDIA GeForce RTX 2070 with Max-Q Design, 7967MiB)
Setup complete ✅ (12 CPUs, 15.4 GB RAM, 536.2/936.8 GB disk)

OS                  Linux-6.8.0-45-generic-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             git
RAM                 15.40 GB
Disk                536.2/936.8 GB
CPU                 Intel Core(TM) i7-10750H 2.60GHz
CPU count           12
GPU                 NVIDIA GeForce RTX 2070 with Max-Q Design, 7967MiB
GPU count           1
CUDA                12.1

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.3.1>=1.8.0
torchvision         ✅ 0.18.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.8>=2.0.0
torch               ✅ 2.3.1!=2.4.0,>=1.8.0; sys_platform == "win32"
```

**Comment by Y-T-G** [comment]
Is `result.boxes.xyxy` empty? Does the saved image have a segmentation mask?

**Comment by rorro6787** [comment]
It does. In fact I made it an image and is all black. And no boxes.xyxy is not empty

**Comment by Y-T-G** [comment]
So the saved image has a mask? Is it a small mask?

You have to multiply the mask with 255 before saving as an image.

**Comment by rorro6787** [comment]
The thing is. It has a mask but is full of zeros. Thats why regardless of multipkying by 255 the result image is a mask but black

**Comment by Y-T-G** [comment]
You should save with `show_boxes=True` and report if the saved box is small or large.

**Comment by Y-T-G** [comment]
But there's no mask annotation in the saved image?

The confusion matrix is based on boxes.

**Comment by Y-T-G** [comment]
You can try increasing the `iou` value to 0.9.

**Comment by JDups** [comment]
I am also receiving masks of all 0's on some predictions that have correct bounding boxes. 

Thing is, I'm encountering this problem running an old project in an environment built from the requirements.txt. It's using yolov8-seg and on ultralytics 8.0.1 and this issue was not happening last year.

I've tried with an ultralytics updated to 8.3.15 and it also happens.

One difference is that this machine has a CUDA version of 12.4 and my previous one has 12.6. Could that be causing this sort of issue?

**Comment by Y-T-G** [comment]
@JDups Can you try increasing `iou` to 0.9?

**Comment by JDups** [comment]
> @JDups Can you try increasing `iou` to 0.9?

I have and it does not have any effect on the problem mask. The are more predictions but the one with an empty mask is still empty.

**Comment by Y-T-G** [comment]
@JDups Is it a new model? Or an old model that didn't have this issue before but has now?

**Comment by JDups** [comment]
@Y-T-G Thank you for your help, while preparing a minimal example I was able to notice the cause of my problems. Some of the pictures I am pulling from my dataset have been slightly changed but have the same filename.

While that explains why a prediction that was successful in the past seemed to no longer work, I still find it unintuitive that a box prediction can be returned with all 0s in the mask. I'm assuming this is due to how thin the boxes are in my cases. The one with all zeroes has a 3.8 width but I have other working masks from 5.2 width boxes.

Thank you again and sorry for my previous confusion.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #19537](https://github.com/ultralytics/ultralytics/pull/19537), authored by @Y-T-G with contributions from @glenn-jocher. 🎉  

### Key Changes in the PR:
The changes aim to optimize YOLO segmentation predictions by introducing a filter to the `construct_result` function that discards predictions with empty masks. This enhancement brings multiple benefits:  
- **Improved Prediction Quality**: Ensures only meaningful segmentation outputs are retained by discarding invalid/empty masks.  
- **Enhanced Efficiency**: Reduces unnecessary handling of irrelevant data, leading to faster performance.  
- **Bug Prevention**: Mitigates potential issues caused by empty masks during downstream processes.  
The version has been incremented from `8.3.83` to `8.3.84` for these updates.  

### Testing the Fix:
Feel free to test the latest changes by installing directly from the repository:  
```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, the changes will be available in the next official PyPI release.  

### Feedback Request:
If possible, please test the updated implementation and let us know if the issue is resolved. Any feedback confirming the fix—or further details if the issue persists—would be greatly appreciated.  

Thank you 🙏 for reporting this issue and contributing to the project. Always happy to hear any additional insights or feedback!

**Comment by glenn-jocher** [comment]
@rorro6787 and @JDups - Thanks for your detailed reports about the segmentation mask issue. This was indeed a bug where the model would return detection boxes but with empty (all zeros) segmentation masks, causing confusion and affecting validation metrics.

Good news! As @UltralyticsAssistant mentioned, the fix has been merged in PR #19537 which specifically filters out predictions with empty masks in the `construct_result` function. This should directly address the issue you both experienced.

The underlying problem was that in some cases (especially with thin objects), the model could detect a bounding box but produce an empty mask, which is inconsistent behavior that affected validation results.

You can try the fix by installing the latest version:
```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Or wait for the next PyPI release (8.3.84+). Please let us know if this resolves your issue after testing.
